src/main/java/com/google/sps/servlets/FavouriteServlet.java

```diff
@@ -0,0 +1,118 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.Recommendation;
+import java.io.IOException;
+import java.lang.Exception;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that sends and gets user's favourite recommendation */
+@WebServlet("/favourite")
+public class FavouriteServlet extends HttpServlet {
+
+  private DatastoreService datastore;
+  private UserService userService;
+
+  @Override
+  public void init(){
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    userService = UserServiceFactory.getUserService();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get list of user's favourite recommendations from datastore
+    String email = userService.getCurrentUser().getEmail();
+    Query query = new Query("Favourite").setFilter(new FilterPredicate("email", FilterOperator.EQUAL, email));
+    PreparedQuery prepared = datastore.prepare(query);
+
+    // Get recommendations based on recommendation ID as key
+    List<Recommendation> favouriteRecommendations = new ArrayList<>();
+    for (Entity favourite : prepared.asIterable()) {
+      long recommendationId = (long) favourite.getProperty("recommendation-id");
+      Key recommendationKey = KeyFactory.createKey("Recommendation", recommendationId);
+      try {
+        Entity recommendation = datastore.get(recommendationKey);
```

Comment:
Reading from a database is usually a very costly operation so it's more efficient to prepare a list with all recommendations keys and query the datastore only once using [`get(java.lang.Iterable<Key> keys)`](http://cloud/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-java.lang.Iterable-)

Reply:
+1